### PR TITLE
Improve `checksum` package and increase its test coverage

### PR DESCRIPTION
**Pull request by cthulhu-rider**

### CODE REVIEW

File: checksum/checksum.go [hunk 1]
```diff
@@ -123,6 +119,7 @@ func Calculate(c *Checksum, t Type, v []byte) {
 	case TZ:
 		c.SetTillichZemor(tz.Sum(v))
 	default:
+		panic(fmt.Errorf("unsupported checksum type %v", t))
```
Comment (carpawell) [inline checksum/checksum.go]: i am not sure it is good. we [have discussed panics](https://github.com/nspcc-dev/neofs-sdk-go/issues/390) in SDK and i thought we do not use them anymore. if a wrong scheme _must be_ (really?) noticed, i would prefer an error

File: checksum/checksum.go [hunk 2]
```diff
@@ -123,13 +123,13 @@ func (c Checksum) WriteToV2(m *refs.Checksum) {
...
 	v2 := (refs.Checksum)(c)
-	switch v2.GetType() {
+	switch typ := v2.GetType(); typ {
 	case refs.SHA256:
 		return SHA256
 	case refs.TillichZemor:
 		return TZ
 	default:
-		return Unknown
+		return Type(typ)
```
Comment (carpawell) [inline checksum/checksum.go:126]: what is motivation? cant it lead to bugs? some code now may operate sums it does not know

**Review by carpawell — COMMENTED** [review]
Conflicts.

File: checksum/checksum.go (hunk 2 above)
Comment (cthulhu-rider) [reply checksum/checksum.go:126]: > what is motivation?

as written in the commit - to not lose the data and work with that we got. Nothing prevents to get real value, not zerofy it

> cant it lead to bugs? some code now may operate sums it does not know

if code is written correctly - no bugs. If incorrectly - the code is already buggy. Good point to detect potential problems

File: checksum/checksum.go (hunk 1 above)
Comment (cthulhu-rider) [reply checksum/checksum.go]: panics are butter than hidden bugs. We discussed this for functions which return an error. Here all we can do is panic. I'll add new function (will try to find an appropriate name) and mark this as deprecated

Comment (carpawell) [reply checksum/checksum.go]: ok, i reopened this issue with some my thoughts: https://github.com/nspcc-dev/neofs-sdk-go/issues/390#issuecomment-2218503588

**Review by carpawell — APPROVED** [review]
We should go further (but conflicts).

**Comment by cthulhu-rider** [comment]
> We should go further (but conflicts).

resolved

**Pull request merged**